src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs

```diff
@@ -62,5 +63,68 @@ public abstract class PipeReader
         /// Cancel the pending <see cref="ReadAsync"/> operation. If there is none, cancels next <see cref="ReadAsync"/> operation, without completing the <see cref="PipeWriter"/>.
         /// </summary>
         public abstract void OnWriterCompleted(Action<Exception, object> callback, object state);
+
+        /// <summary>
+        /// Asynchronously reads the bytes from the <see cref="PipeReader"/> and writes them to the specified stream, using a specified buffer size and cancellation token.
+        /// </summary>
+        /// <param name="destination">The stream to which the contents of the current stream will be copied.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.</param>
+        /// <returns>A task that represents the asynchronous copy operation.</returns>
+        public virtual async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+        {
+            while (!cancellationToken.IsCancellationRequested)
```

Comment:
I'm not sure this is actually what you want.  If cancellation is requested and hits this check, the method will just end, whereas if it's requested between here and the ReadAsync, it would likely result in a cancellation exception getting thrown.  Seems like this should just be `while (true)`.

Reply:
Yes, thats why I had that check at the bottom, but maybe a while true is fine.

Reply:
> Yes, thats why I had that check at the bottom

That's then a race condition: you could do the check at the bottom of the loop, when cancellation wasn't requested, and by the time you do the check at the top of the loop, it has been requested.

Reply:
Yes, I'm going back to while true.